test-data/unit/fine-grained.test

```diff
@@ -1797,6 +1797,372 @@ def foo(x: Point) -> int:
+[builtins fixtures/dict.pyi]
+[out]
+==
+b.py:2: error: Dict entry 0 has incompatible type "str": "int"; expected "str": "str"
+
+[case testAliasFineNestedFunc]
+import b
+[file a.py]
+from typing import Dict
+A = Dict[str, int]
+[file a.py.2]
+from typing import Dict
+A = Dict[str, str]
+[file aa.py]
+from typing import Dict
+import a
+B = Dict[str, a.A]
+[file b.py]
+import aa
+def f(x: aa.B):
+    x = {'first': {str(): int()}}
+[builtins fixtures/dict.pyi]
+[out]
+==
+b.py:3: error: Dict entry 0 has incompatible type "str": "int"; expected "str": "str"
+
+[case testAliasFineNestedFuncDirect]
+import b
+[file a.py]
+from typing import Dict
+A = Dict[str, int]
+[file a.py.2]
+from typing import Dict
+A = Dict[str, str]
+[file aa.py]
+from typing import Dict
+import a
+E = Dict
+[file b.py]
+import aa
+def f(x: aa.E[str, aa.a.A]):
+    x = {'first': {str(): int()}}
+[builtins fixtures/dict.pyi]
+[out]
+==
+b.py:3: error: Dict entry 0 has incompatible type "str": "int"; expected "str": "str"
+
+[case testAliasFineNonGenericToGeneric]
+import b
+[file a.py]
+from typing import Dict, TypeVar
+T = TypeVar('T')
+A = Dict[T, int]
+[file a.py.2]
+A = str
+[file b.py]
+import a
+def f(x: a.A[str]):
+    pass
+[builtins fixtures/dict.pyi]
+[out]
+==
+b.py:2: error: "str" expects no type arguments, but 1 given
+
+[case testAliasFineGenericToNonGeneric]
+import b
+[file a.py]
+A = str
+[file a.py.2]
+from typing import Dict, TypeVar
+T = TypeVar('T')
+A = Dict[T, int]
+[file b.py]
+import a
+def f(x: a.A):
+    pass
+reveal_type(f)
+[builtins fixtures/dict.pyi]
+[out]
+b.py:4: error: Revealed type is 'def (x: builtins.str) -> Any'
+==
+b.py:4: error: Revealed type is 'def (x: builtins.dict[Any, builtins.int]) -> Any'
+
+[case testAliasFineChangedNumberOfTypeVars]
+import b
+[file a.py]
+from typing import Dict, TypeVar
+T = TypeVar('T')
+A = Dict[T, int]
+[file a.py.2]
+from typing import Dict, TypeVar
+T = TypeVar('T')
+S = TypeVar('S')
+A = Dict[T, S]
+[file b.py]
+import a
+def f(x: a.A[str]):
+    pass
+[builtins fixtures/dict.pyi]
+[out]
+==
+b.py:2: error: Bad number of arguments for type alias, expected: 2, given: 1
+
+[case testAliasFineComponentDeleted]
+import b
+[file a.py]
+class B: pass
+[file a.py.2]
+x = 1
+[file b.py]
+import a
+from typing import Dict, TypeVar
+T = TypeVar('T')
+A = Dict[T, a.B]
+def f(x: A[int]):
+    pass
+[builtins fixtures/dict.pyi]
+[out]
+==
+b.py:4: error: Name 'a.B' is not defined
+
+[case testAliasFineTargetDeleted]
+import c
+[file a.py]
+A = int
+[file b.py]
+import a
+B = a.A
+[file b.py.2]
+x = 1
+[file c.py]
+import b
+def f(x: b.B):
+    pass
+[out]
+==
+c.py:2: error: Name 'b.B' is not defined
+
+[case testAliasFineClassInFunction]
+import b
+[file a.py]
+A = int
+[file a.py.2]
+A = str
+[file b.py]
+import a
+def f() -> None:
+    class C:
+        x: a.A = int()
+[out]
+==
+b.py:4: error: Incompatible types in assignment (expression has type "int", variable has type "str")
```

Comment:
Potential new tests (if they don't already exist):

* Create a type alias (initially there is a stale reference to a type alias).
* Delete type alias (after update there is a stale reference to a type alias).
* Switch a class to a type alias (or type alias to class).